starlette/applications.py

```diff
@@ -124,10 +127,15 @@ def mount(self, path: str, app: ASGIApp, name: str | None = None) -> None:
     def host(self, host: str, app: ASGIApp, name: str | None = None) -> None:
         self.router.host(host, app=app, name=name)  # pragma: no cover
 
-    def add_middleware(self, middleware_class: type, **options: typing.Any) -> None:
+    def add_middleware(
+        self,
+        middleware_class: typing.Type[MiddlewareClass[P]],
```

Comment:
I figured that we could achieve more precise annotation via `typing.Protocol`. Moreover, this ensures that the middleware instance does adhere to the ASGI protocol.